File: app/src/main/java/com/highcom/comicmemo/ListDataManager.java
package com.highcom.comicmemo;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.Collections;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ListDataManager {
    private static ListDataManager listDataManager;
    private SQLiteDatabase rdb;
    private SQLiteDatabase wdb;
    private Map<String, String> data;
    private List<Map<String, String>> dataList1;
    private List<Map<String, String>> dataList2;
    private String sortKey1;
    private String sortKey2;
    private int lastUpdateId;

    public static ListDataManager createInstance(Context context) {
        listDataManager = new ListDataManager(context);
        return listDataManager;
    }

    public static ListDataManager getInstance() {
        return listDataManager;
    }

    private ListDataManager(Context context) {
        lastUpdateId = 0;
        sortKey1 = "id";
        sortKey2 = "id";
        ListDataOpenHelper helper = new ListDataOpenHelper(context);
        rdb = helper.getReadableDatabase();
        wdb = helper.getWritableDatabase();
        dataList1 = new ArrayList<Map<String, String>>();
        dataList2 = new ArrayList<Map<String, String>>();
        remakeAllListData();
    }

    public void setLastUpdateId(int id) {
        lastUpdateId = id;
    }

    public int getLastUpdateId() {
        return lastUpdateId;
    }

    public void setData(boolean isEdit, Map<String, String> data) {
        // データベースに追加or編集する
        ContentValues values = new ContentValues();
        values.put("id", Long.valueOf(data.get("id")).longValue());
        values.put("title", data.get("title"));
        values.put("author", data.get("author"));
        values.put("number", data.get("number"));
        values.put("memo", data.get("memo"));
        values.put("inputdate", data.get("inputdate"));
        values.put("status", data.get("status"));
        if (isEdit) {
            // 編集の場合
            wdb.update("comicdata", values, "id=?", new String[] { data.get("id") });
        } else {
            // 新規作成の場合
            wdb.insert("comicdata", data.get("id"), values);
        }
        remakeAllListData();
    }

    public void deleteData(String id) {
        // データベースから削除する
        wdb.delete("comicdata", "id=?", new String[] { id });

        remakeAllListData();
    }

    public List<Map<String, String>> getDataList(long dataIndex) {
        List<Map<String, String>> dataList;
        if (dataIndex == 0) {
            dataList = dataList1;
        } else {
            dataList = dataList2;
        }
        return dataList;
    }

    public void rearrangeData(long dataIndex, int fromPos, int toPos) {
        boolean mov;
        ContentValues values;

        Cursor cur = getCurrentCursor(dataIndex);

        mov = cur.moveToPosition(fromPos);
        if (!mov) return;
        long fromId = cur.getLong(0);

        mov = cur.moveToPosition(toPos);
        if (!mov) return;
        long toId = cur.getLong(0);

        values = new ContentValues();
        values.put("id", -1);
        wdb.update("comicdata", values, "id=?", new String[] { Long.toString(fromId) });

        values = new ContentValues();
        values.put("id", fromId);
        wdb.update("comicdata", values, "id=?", new String[] { Long.toString(toId) });

        values = new ContentValues();
        values.put("id", toId);
        wdb.update("comicdata", values, "id=?", new String[] { Long.toString(-1) });

        remakeAllListData();
    }

    public long getNewId() {
        long newId = 0;
        Cursor cur = getAllCursor();

        boolean mov = cur.moveToFirst();
        long curId;
        while (mov) {
            curId = Long.valueOf(cur.getString(0)).longValue();
            if (newId < curId) {
                newId = curId;
            }
            mov = cur.moveToNext();
        }

        return newId + 1;
    }

    public void closeData() {
        rdb.close();
        wdb.close();
    }

    public void remakeAllListData() {
        remakeListData(0);
        sortListData(0, sortKey1);
        remakeListData(1);
        sortListData(1, sortKey2);
    }

    public void remakeListData(long dataIndex) {
        List<Map<String, String>> dataList;
        if (dataIndex == 0) {
            dataList = dataList1;
        } else {
            dataList = dataList2;
        }
        dataList.clear();

        Cursor cur = getCurrentCursor(dataIndex);

        boolean mov = cur.moveToFirst();
        while (mov) {
            data = new HashMap<String, String>();
            data.put("id", cur.getString(0));
            data.put("title", cur.getString(1));
            data.put("author", cur.getString(2));
            data.put("number", cur.getString(3));
            data.put("memo", cur.getString(4));
            data.put("inputdate", cur.getString(5));
            data.put("status", cur.getString(6));
            dataList.add(data);
            mov = cur.moveToNext();
        }
    }

    public void sortListData(long dataIndex, final String key) {
        List<Map<String, String>> dataList;
        if (dataIndex == 0) {
            dataList = dataList1;
            sortKey1 = key;
        } else {
            dataList = dataList2;
            sortKey2 = key;
        }
        Collections.sort(dataList, new Comparator<Map<String, String>>() {
            @Override
            public int compare(Map<String, String> stringStringMap, Map<String, String> t1) {
                int result;
                if (key == "id") {
                   result = Integer.valueOf(stringStringMap.get("id")).compareTo(Integer.valueOf(t1.get("id")));
                } else {
                    result = stringStringMap.get(key).compareTo(t1.get(key));
                }

                // ソート順が決まらない場合には、idで比較する
                if (result == 0) {
                    result = Integer.valueOf(stringStringMap.get("id")).compareTo(Integer.valueOf(t1.get("id")));
                }
                return result;
            }
        });
    }

    private Cursor getCurrentCursor(long dataIndex) {
        return rdb.query("comicdata", new String[] { "id", "title", "author", "number", "memo", "inputdate", "status" }, "status=?", new String[]{Long.toString(dataIndex)}, null, null, "id ASC");
    }

    private Cursor getAllCursor() {
        return rdb.query("comicdata", new String[] { "id", "title", "author", "number", "memo", "inputdate", "status" }, null, null, null, null, "id ASC");
    }
}

File: app/src/main/java/com/highcom/comicmemo/ListDataManager.kt
package com.highcom.comicmemo

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import java.util.*

/**
 * 巻数の一覧データ管理クラス
 *
 * @param context コンテキスト
 */
class ListDataManager private constructor(context: Context) {
    /** 読み取り用DBアクセス */
    private val rdb: SQLiteDatabase
    /** 書き込み用DBアクセス */
    private val wdb: SQLiteDatabase
    /** 巻数データ */
    private var data: MutableMap<String, String>? = null
    /** 続刊用のデータリスト */
    private val dataList1: MutableList<Map<String, String>>
    /** 完結用のデータリスト */
    private val dataList2: MutableList<Map<String, String>>
    /** 続刊用のソートキー */
    private var sortKey1 = "id"
    /** 完結用のソートキー */
    private var sortKey2 = "id"
    /** 最後に更新した巻数データID */
    var lastUpdateId = 0

    /**
     * 作成した巻数データの登録処理
     *
     * @param isEdit 編集かどうか
     * @param data 巻数データ
     */
    fun setData(isEdit: Boolean, data: Map<String, String>) {
        // データベースに追加or編集する
        val values = ContentValues()
        values.put("id", java.lang.Long.valueOf(data["id"]).toLong())
        values.put("title", data["title"])
        values.put("author", data["author"])
        values.put("number", data["number"])
        values.put("memo", data["memo"])
        values.put("inputdate", data["inputdate"])
        values.put("status", data["status"])
        if (isEdit) {
            // 編集の場合
            wdb.update("comicdata", values, "id=?", arrayOf(data["id"]))
        } else {
            // 新規作成の場合
            wdb.insert("comicdata", data["id"], values)
        }
        remakeAllListData()
    }

    /**
     * 巻数データ削除処理
     *
     * @param id 削除対象のID
     */
    fun deleteData(id: String?) {
        // データベースから削除する
        wdb.delete("comicdata", "id=?", arrayOf(id))
        remakeAllListData()
    }

    /**
     * 巻数データ一覧取得処理
     *
     * @param dataIndex 巻数リストデータインデックス 0:続刊 1:完結
     * @return 巻数データ一覧
     */
    fun getDataList(dataIndex: Long): List<Map<String, String>> {
        val dataList: List<Map<String, String>> = if (dataIndex == 0L) {
            dataList1
        } else {
            dataList2
        }
        return dataList
    }

    /**
     * 巻数データの入れ替え処理
     *
     * @param dataIndex 巻数リストデータインデックス 0:続刊 1:完結
     * @param fromPos 入れ替え元ID
     * @param toPos 入れ替え先ID
     */
    fun rearrangeData(dataIndex: Long, fromPos: Int, toPos: Int) {
        var mov: Boolean
        val cur = getCurrentCursor(dataIndex)
        mov = cur.moveToPosition(fromPos)
        if (!mov) return
        val fromId = cur.getLong(0)
        mov = cur.moveToPosition(toPos)
        if (!mov) return
        val toId = cur.getLong(0)
        var values = ContentValues()
        values.put("id", -1)
        wdb.update("comicdata", values, "id=?", arrayOf(java.lang.Long.toString(fromId)))
        values = ContentValues()
        values.put("id", fromId)
        wdb.update("comicdata", values, "id=?", arrayOf(java.lang.Long.toString(toId)))
        values = ContentValues()
        values.put("id", toId)
        wdb.update("comicdata", values, "id=?", arrayOf(java.lang.Long.toString(-1)))
        remakeAllListData()
    }

    /**
     * 新規IDの発行プロパティ
     */
    val newId: Long
        get() {
            var newId: Long = 0
            val cur = allCursor
            var mov = cur.moveToFirst()
            var curId: Long
            while (mov) {
                curId = java.lang.Long.valueOf(cur.getString(0)).toLong()
                if (newId < curId) {
                    newId = curId
                }
                mov = cur.moveToNext()
            }
            return newId + 1
        }

    /**
     * データベースをクローズする
     *
     */
    fun closeData() {
        rdb.close()
        wdb.close()
    }

    /**
     * 巻数データ一覧の再作成処理
     *
     */
    private fun remakeAllListData() {
        remakeListData(0)
        sortListData(0, sortKey1)
        remakeListData(1)
        sortListData(1, sortKey2)
    }

    /**
     * 指定された巻数データ一覧の再作成処理
     *
     * @param dataIndex 巻数リストデータインデックス 0:続刊 1:完結
     */
    private fun remakeListData(dataIndex: Long) {
        val dataList: MutableList<Map<String, String>> = if (dataIndex == 0L) {
            dataList1
        } else {
            dataList2
        }
        dataList.clear()
        val cur = getCurrentCursor(dataIndex)
        var mov = cur.moveToFirst()
        while (mov) {
            data = HashMap()
            (data as HashMap<String, String>)["id"] = cur.getString(0)
            (data as HashMap<String, String>)["title"] = cur.getString(1)
            (data as HashMap<String, String>)["author"] = cur.getString(2)
            (data as HashMap<String, String>)["number"] = cur.getString(3)
            (data as HashMap<String, String>)["memo"] = cur.getString(4)
            (data as HashMap<String, String>)["inputdate"] = cur.getString(5)
            (data as HashMap<String, String>)["status"] = cur.getString(6)
            dataList.add(data as HashMap<String, String>)
            mov = cur.moveToNext()
        }
    }

    /**
     * 指定された巻数データ一覧を指定されたキーでソートする
     *
     * @param dataIndex 巻数リストデータインデックス 0:続刊 1:完結
     * @param key ソートキー
     */
    fun sortListData(dataIndex: Long, key: String) {
        val dataList: List<Map<String, String>>
        if (dataIndex == 0L) {
            dataList = dataList1
            sortKey1 = key
        } else {
            dataList = dataList2
            sortKey2 = key
        }

        val comparator = Comparator<Map<String, String>> { t1, t2 ->
            var result = if (key === "id") {
                t1["id"]!!.compareTo(t2["id"]!!)
            } else {
                t1[key]!!.compareTo(t2[key]!!)
            }

            // ソート順が決まらない場合には、idで比較する
            if (result == 0) {
                result = t1["id"]!!.compareTo(t2["id"]!!)
            }
            return@Comparator result
        }
        dataList.sortedWith(comparator)
    }

    /**
     * 現在のDBカーソル取得処理
     *
     * @param dataIndex 巻数リストデータインデックス 0:続刊 1:完結
     * @return カーソル
     */
    private fun getCurrentCursor(dataIndex: Long): Cursor {
        return rdb.query(
            "comicdata",
            arrayOf("id", "title", "author", "number", "memo", "inputdate", "status"),
            "status=?",
            arrayOf(java.lang.Long.toString(dataIndex)),
            null,
            null,
            "id ASC"
        )
    }

    /**
     * 全データのカーソルプロパティ
     */
    private val allCursor: Cursor
        get() = rdb.query(
            "comicdata",
            arrayOf("id", "title", "author", "number", "memo", "inputdate", "status"),
            null,
            null,
            null,
            null,
            "id ASC"
        )

    companion object {
        var instance: ListDataManager? = null
            private set

        fun createInstance(context: Context): ListDataManager? {
            instance = ListDataManager(context)
            return instance
        }
    }

    init {
        val helper = ListDataOpenHelper(context)
        rdb = helper.readableDatabase
        wdb = helper.writableDatabase
        dataList1 = ArrayList()
        dataList2 = ArrayList()
        remakeAllListData()
    }
}